File: week1/homework1.py
import string
def better_solution(random_word,dictionary):
    sorted_random_word="".join(sorted(random_word))
    new_dictionary=[]
    for word in dictionary:
        sorted_word="".join(sorted(word))
        new_dictionary.append((sorted_word,word))
    new_dictionary.sort()
    anagrams=[]
    target=(sorted_random_word, "")
    
    left=0
    right=len(new_dictionary)
    while left<right:
        mid=(left+right)//2
        if new_dictionary[mid]<target:
            left=mid+1
        else:
            right=mid
    index=left
    
    while (index<len(new_dictionary) and new_dictionary[index][0]==sorted_random_word):
        if new_dictionary[index][1]!=random_word:
            anagrams.append(new_dictionary[index][1])
        index+=1
    return anagrams

dictionary = []
with open('words.txt','r',encoding='utf-8') as file:
    for line in file:
        word = line.strip()
        if word != "":
            dictionary.append(word)

random_word = input("アナグラムを探したい英単語を入力してください: ")

result = better_solution(random_word, dictionary)

if len(result) == 0:
    print("アナグラムは見つかりませんでした!")
else:
    for ans in result:
        print(ans)
        
        
"""実行結果
yanagisawa.yukia@YukiAir week1 % python3 homework1.py
アナグラムを探したい英単語を入力してください: tea
ate
eat
eta
tae
yanagisawa.yukia@YukiAir week1 % python3 homework1.py
アナグラムを探したい英単語を入力してください: xyz
アナグラムは見つかりませんでした。"""

File: week1/homework1.cpp
#include <bits/stdc++.h>
using namespace std;
int main() {
    ifstream input("words.txt");
    if(!input) {
        cerr << "ファイルを開くのに失敗しました。" << endl;
        return 1;
    }
    vector<pair<string,string>> new_dictionary;
    string word;
    while(input >> word) {
        string sorted_word = word;
        sort(sorted_word.begin(),sorted_word.end());
        new_dictionary.push_back({sorted_word,word});
    }
    input.close();

    sort(new_dictionary.begin(),new_dictionary.end());
    string random_word;
    cout << "アナグラムを探したい文字列：";
    getline(cin, random_word);
    string sorted_random_word = random_word;
    sort(sorted_random_word.begin(),sorted_random_word.end());

    bool found = false;
    auto it=lower_bound(
        new_dictionary.begin(),
        new_dictionary.end(),
        sorted_random_word,
        [](const pair<string, string>& p, const string& value) {
            return p.first<value;
        }
    );

    while(it!=new_dictionary.end()&&it->first==sorted_random_word) {
        if(it->second!=random_word) {
            cout << it->second << endl;
            found=true;
        }
        it++;
    }
    if(!found) {
        cout << "アナグラムは見つかりませんでした！" << endl;
    }

    return 0;
}

/*実行結果
yanagisawa.yukia@YukiAir week1 % ./homework1                      
アナグラムを探したい文字列：listen
enlist
inlets
silent
tinsel
yanagisawa.yukia@YukiAir week1 % ./homework1
アナグラムを探したい文字列：xyz
アナグラムは見つかりませんでした！
*/